### PR TITLE
Mean now handle iteratively instead of summing to overflow

### PR DESCRIPTION
**Pull request by ytyubox**
I recently got this issue for summing may overflow. This may cause fatal. I just add the functionality randomly, maybe you want to rearrange files again.

### CODE REVIEW

**Comment by ytyubox** [comment]
No idea why the build failure, maybe `Swift 5.2` causes the problem?

**Comment by JimmyMAndersson** [comment]
@ytyubox Very nice catch! I'll have a look at the build failure and any file rearrangements when I get to my workstation this afternoon.

File: Sources/StatKit/Descriptive Statistics/Central Tendency/MeanType.swift [hunk 1]
```diff
@@ -25,7 +25,7 @@ extension MeanType {
                                                              in collection: U) -> Double {
     switch self {
       case .arithmetic:
-        return collection.sum(over: variable).realValue / Double(collection.count)
+        return collection.average(over: variable)
```
Comment (JimmyMAndersson) [inline Sources/StatKit/Descriptive Statistics/Central Tendency/MeanType.swift:28]: My suggestion is to inline the calculation part that resided in the Summation file, so the code that computes the mean is all in the same place.

```suggestion
        return zip(collection, 1...collection.count)
          .reduce(into: 0.0) { result, zip in
            let (element, count) = zip
            result += (element[keyPath: variable].realValue - result) / count.realValue
      }
```

File: Sources/StatKit/Descriptive Statistics/Summation/Summation.swift [hunk 2]
```diff
@@ -10,4 +10,21 @@ extension Sequence {
...
+    {
+        var avg = 0.0
+        var t = 1.0
+          for x in self {
+            let value = x[keyPath: variable]
+            avg += (value.realValue - avg) / t
+            t += 1
+          }
+          return avg
+    }
```
Comment (JimmyMAndersson) [inline Sources/StatKit/Descriptive Statistics/Summation/Summation.swift:29]: I think we can move this computation into MeanType, as mentioned in the above comment.

```suggestion
```

File: Tests/StatKitTests/Descriptive Statistics Tests/AveragesTests.swift [hunk 3]
```diff
@@ -11,6 +11,14 @@ final class AveragesTests: XCTestCase {
...
     XCTAssertEqual(calculatedMean, expectedMean)
   }
+  func testBigNumberIntegerArithmeticMean() {
+    let value = Int.max / 9
+    let bigIntArray = [9*value, 6*value, 3*value]
+    let calculateMean = bigIntArray.mean(of: \.self)
+    let expectedMean = Double(value*6)
+    print(expectedMean, calculateMean, bigIntArray)
+    XCTAssertEqual(calculateMean, expectedMean)
+  }
```
Comment (JimmyMAndersson) [inline Tests/StatKitTests/Descriptive Statistics Tests/AveragesTests.swift:21]: The debug print should be taken out. :)

```suggestion

  func testBigNumberIntegerArithmeticMean() {
    let value = Int.max / 9
    let bigIntArray = [9 * value, 6 * value, 3 * value]
    let calculateMean = bigIntArray.mean(of: \.self)
    let expectedMean = Double(value * 6)
    
    XCTAssertEqual(calculateMean, expectedMean, accuracy: 0.00001)
  }
```

**Comment by JimmyMAndersson** [comment]
Btw, that way of calculating the mean you're using is really nifty. I like it!

File: Sources/StatKit/Descriptive Statistics/Central Tendency/MeanType.swift (hunk 1 above)
Comment (ytyubox) [reply Sources/StatKit/Descriptive Statistics/Central Tendency/MeanType.swift:28]: I was trying to use `enumerate()` to replace `zip()`, but it looks like a mess. 
```Swift
collection.enumerated().reduce(into: 0.0) {
            (result: inout Double, pair: EnumeratedSequence<U>.Element)  in
            let (index, element) = pair
            let count = index + 1  // terrible magic number `+1`
            result += (element[keyPath: variable].realValue - result) / count.realValue
        }
```
I think zip is a great choice!

**Comment by ytyubox** [comment]
This is my first time using GitHub's `Changes requested`, no idea how to solve the blocking for merge.

**Comment by JimmyMAndersson** [comment]
The merge block is just that a pull request requires an approved review before it can be merged. It's been approved and merged into master by a squash merge, so you'll find the changes in the latest commit.

Thanks for contributing! :)

**Comment by ytyubox** [comment]
Glad that I can help! BTW I love that `Co-authored-by` commit. 😄

**Comment by JimmyMAndersson** [comment]
I didn't read the commit message until now, but those may be automatically generated by the suggestions in the first review, I guess. 😅

**Pull request merged**